dxp/resource_injection.md

```diff
@@ -0,0 +1,106 @@
+
+### Declarative Syntax
+
+The declarative syntax is used in `bnd.bnd` files and makes use of three different headers:
+
+-   `Liferay-JS-Resources-Top-Head`: this specifies a comma delimited list of JavaScript resources to include always. The path of the files is relative to `src/main/resource/META-INF/resources` as they are fetched through the bundle's servlet context.
+-   `Liferay-JS-Resources-Top-Head-Authenticated`: this specifies a comma delimited list of JavaScript resources to include **only when the user has been authenticated**. The path of the files is relative to `src/main/resources/META-INF/resources` as they are fetched through the bundle's servlet context.
+-   `Liferay-Top-Head-Weight`: the value of this property is used as [`service.ranking`](https://docs.osgi.org/javadoc/r2/org/osgi/framework/Constants.html#SERVICE_RANKING) to prioritize the associated `TopHeadResources`.
+
+Whenever a bundle containing one of the first two headers is deployed to DXP, [TopHeadExtender](https://github.com/liferay/liferay-portal/blob/b8db59dbc4a75e6bdc38d01d7c54325ca6425c88/modules/apps/frontend-js/frontend-js-top-head-extender/src/main/java/com/liferay/frontend/js/top/head/extender/internal/TopHeadExtender.java) scans it and creates a `TopHeadResources` on the fly.
+
+# `liferay-util:html-top`, `liferay-util:html-bottom`, and `liferay-util:body-bottom` tags
+
+The [`<html-top>`](https://github.com/liferay/liferay-portal/blob/b87113b5cfe9217b8327ce60c97579938c7fc6dd/util-taglib/src/com/liferay/taglib/util/HtmlTopTag.java), [`<html-bottom>`](https://github.com/liferay/liferay-portal/blob/b87113b5cfe9217b8327ce60c97579938c7fc6dd/util-taglib/src/com/liferay/taglib/util/HtmlBottomTag.java) and [`<body-bottom>`](https://github.com/liferay/liferay-portal/blob/b87113b5cfe9217b8327ce60c97579938c7fc6dd/util-taglib/src/com/liferay/taglib/util/BodyBottomTag.java) tags can be used to inject arbitrary HTML code to the rendered page at specific positions:
+
+-   [`<html-top>` is injected at `themes/top_head.jsp`](https://github.com/liferay/liferay-portal/blob/e3ffac158e0ec5acc5c67069fbd7ba688d3c78d4/portal-web/docroot/html/common/themes/top_head.jsp#L125)
+-   [`<html-bottom>` is injected at `themes/bottom.jsp`](https://github.com/liferay/liferay-portal/blob/e3ffac158e0ec5acc5c67069fbd7ba688d3c78d4/portal-web/docroot/html/common/themes/bottom.jsp#L52)
+-   [`<body-bottom>` is injected at `themes/body_bottom.jsp`](https://github.com/liferay/liferay-portal/blob/a009a8e65f27d8c0640cd44ba661ff12994649f2/portal-web/docroot/html/common/themes/body_bottom.jsp#L20)
```

Comment:
Is this the order that they appear on the page? An explanatory label would help to distinguish between "bottom" and "body_bottom" especially?

Reply:
AFAIK), these insertion points are somewhat arbitrary, and it's not clear enough why we have these and not others, or more, or less... :shrug: 

It looks like these are key points in the page rendering lifecycle, but they don't seem to be carefully defined and documented. 

For example: I would expect something like "this is the insertion point you need to use if you want to inject some JavaScript files before the AMD architecture has been initialized" or "this is where you inject CSS files before the theme's CSS files are injected". That is: each key should explain what is available at that point and what will be loaded afterwards.

Instead, we have keys referring to the place where you will insert your content, without any clue as to what it is exactly. So, the only way to know what will happen is to try it and see if it fits what you want to do.

Perhaps @jbalsas has any _secret_ knowledge about this he wants to share? :thinking:

Reply:
There's no _secret_ knowledge, just lack of documentation. We started documenting some but I think that effort got discontinued. You can read about some of these in this [Dynamic Includes
 article](https://help.liferay.com/hc/en-us/articles/360018165711-Dynamic-Includes).

While we're at it, it would be good to try and do some initial exploration to explain the different extension points so the docs team can later translate this to developer guides.

Reply:
That article looks like what I was asking for :+1: .

Regarding the documentation of dynamic includes, it's OK, we can document them, but IMO it would be even better if we defined good coherent APIs (like we have done with `TopHeadResources`) that can be documented, extended, and better modeled than a key + `DynamicInclude`.

Reply:
Shall we merge this? I added the link you posted @jbalsas , but I don't think it makes sense to stop this from merging until we analyze existing dynamic includes. It may take a long time and it's not clear if they are still used in an isolated way that can be documented. I mean, people tend to put everything in top_head and only move it somewhere else if it doesn't work.

As I said, it would be much better if we define proper APIs for proper points in the render lifecycle and make people use them. The dynamic include was a proposal for an extension mechanism based on what we had by that time (a bit better than hooks but no more than that), but it was never meant to be used as a serious high level solution for everybody/every use case.

Reply:
Go ahead, in this case, I think more is better, no need to wait for my confirmation.